### PR TITLE
[IMP] stock: Add hybrid procure method 'mts_else_mto'

### PR DESCRIPTION
**Pull request by arbaes**
In some occasions, products can be available in a location, but are still ignored when the MTO is set on those products.
(e.g.: When a product is returned)

To avoid this, we add a `procure_method` named `mts_else_mto`, allowing procurement to reserve the products if enough of them are available in his source location.
In order to achieve this,  we apply the MTS or the MTO `procure_method` on the moves from the procurement depending on whether or not there are enough quantities of the needed products in the source location ('All or nothing' logic).

TaskID: 1831347

### CODE REVIEW

File: addons/sale_stock/tests/test_sale_stock.py [hunk 1]
```diff
@@ -497,4 +497,96 @@ def test_07_forced_qties(self):
         })
         so1.picking_ids.button_validate()
         self.assertEqual(so1.picking_ids.state, 'done')
-        self.assertEqual(so1.order_line.mapped('qty_delivered'), [1, 1, 1])
\ No newline at end of file
+        self.assertEqual(so1.order_line.mapped('qty_delivered'), [1, 1, 1])
+
+    def test_mtso_mto(self):
```
Comment (sle-odoo) [inline addons/sale_stock/tests/test_sale_stock.py]: move to stock in TestRoutes

File: addons/sale_stock/tests/test_sale_stock.py [hunk 2]
```diff
@@ -497,4 +497,96 @@ def test_07_forced_qties(self):
...
+        # 3 pickings shloud be created.
+        self.assertEquals(len(so.picking_ids), 3)
+        for picking in so.picking_ids:
+            # Only the picking from Stock to Pack should be MTS
+            if picking.location_id == warehouse.lot_stock_id:
+                self.assertEquals(picking.move_lines.procure_method, 'make_to_stock')
+            else:
+                self.assertEquals(picking.move_lines.procure_method, 'make_to_order')
+
+            # TODO: To update if we change the 'all or nothing' mechanism
```
Comment (sle-odoo) [inline addons/sale_stock/tests/test_sale_stock.py]: remove

File: addons/stock/models/stock_rule.py [hunk 3]
```diff
@@ -44,10 +44,13 @@ class StockRule(models.Model):
     route_id = fields.Many2one('stock.location.route', 'Route', required=True, ondelete='cascade')
     procure_method = fields.Selection([
         ('make_to_stock', 'Take From Stock'),
-        ('make_to_order', 'Trigger Another Rule')], string='Move Supply Method',
+        ('make_to_order', 'Trigger Another Rule'),
+        ('mts_then_mto', 'Take From Stock, if Unavailable, Trigger Another Rule')], string='Move Supply Method',
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: weird capitals

File: addons/stock/models/stock_rule.py [hunk 4]
```diff
@@ -167,6 +172,21 @@ def _push_prepare_move_copy_values(self, move_to_copy, new_date):
         }
         return new_move_vals
 
+    def _get_procure_method(self, product, qty_needed, location):
+        """ When the procure_method of the rule is 'mts_then_mto', we need to get the quantity available to know if
+        we have to apply MTS or MTO on the moves of the procurement.
+        :param product: The product of the move
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: newline

File: addons/stock/models/stock_rule.py [hunk 5]
```diff
@@ -167,6 +172,21 @@ def _push_prepare_move_copy_values(self, move_to_copy, new_date):
         }
         return new_move_vals
 
+    def _get_procure_method(self, product, qty_needed, location):
+        """ When the procure_method of the rule is 'mts_then_mto', we need to get the quantity available to know if
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: quantity forecasted?

File: addons/stock/models/stock_rule.py [hunk 6]
```diff
@@ -167,6 +172,21 @@ def _push_prepare_move_copy_values(self, move_to_copy, new_date):
...
+        we have to apply MTS or MTO on the moves of the procurement.
+        :param product: The product of the move
+        :param qty_needed: The amount needed of that product
+        :param location: The location source of the product
+        :return: The procure_method to apply on the move
+        """
+        if self.procure_method == 'mts_then_mto':
+            qty_available = product.with_context(location=location.id).virtual_available
```
Comment (sle-odoo) [inline addons/stock/models/stock_rule.py]: can't you batch virtual_available? read it for all products where the rule is mts_then_mto and pass the values there

**Comment by rvalyi** [comment]
Hello, great news! Eventually you will want to check it covers the same features as this OCA module by Akretion we had since Odoo 8 https://github.com/OCA/stock-logistics-warehouse/tree/11.0/stock_mts_mto_rule

**Comment by blaggacao** [comment]
Standing ovations!

**Comment by arbaes** [comment]
Will be replaced by #32778

**Pull request closed without merging**